### PR TITLE
Add_Process_Reward

### PR DESCRIPTION
**Pull request by myAugust**
### **Add_process_reward**

Considering that tool invocation requires multi-turn interactions, we aim to investigate whether introducing **process rewards** can help the model learn more effectively and efficiently. 

Based on this motivation, the following modifications have been implemented:

1. **Added Stop Tokens with Process Rewards**  
   - Assure a stop symbol at each tool invocation step, accompanied by a process reward for that step.  

2. **Configurable Process Rewards**  
   - Users can enable/disable process rewards via a script parameter:  
     ```yaml
     actor_rollout_ref.env.use_process_reward: False  # or True
     ```  
   - Ensures flexibility for ablation studies.  

3. **Extensible Reward Design**  
   - Currently implemented simple rule-based rewards (e.g.,reward based on whether the tool invocation format is correct).  
   - Future extensions: Support diverse reward forms (e.g., LLM-generated feedback).  

--- 

### **Key Focus**  
- Validate if process rewards reduce "reward sparsity" in multi-turn tool-calling scenarios.

### CODE REVIEW

File: envs/utils/tool_utils.py [hunk 1]
```diff
@@ -46,21 +48,32 @@ def postprocess_output(self, output: DataProto, step: int):
             batch_idxs = output.meta_info['index']
 
         responses = output.batch.get('responses')
+        responses_str=[]
```
Comment (James0618) [inline envs/utils/tool_utils.py:51]: if self.env_object.use_process_reward放在前面吧，如果用再做for循环，不然就还是batch_decode

File: envs/utils/tool_utils.py [hunk 2]
```diff
@@ -146,6 +160,21 @@ def compose_final_output(self, step) -> DataProto:
         response_token = torch.tensor(input_ids_list, dtype=torch.int64)[:,:max_len]
         response_loss_mask = torch.tensor(loss_mask_list, dtype=torch.float32)
         response_attention_mask = (response_token != self.pad_token_id).long()
+
+
+        #max_tool_use_len = self.max_turns
```
Comment (James0618) [inline envs/utils/tool_utils.py:165]: 加一点注释

File: verl/trainer/main_ppo.py [hunk 3]
```diff
@@ -182,13 +182,18 @@ def run(self, config):
             reward_fn.set_env_object(env_object)
         except Exception as e:
             print(f"Error setting env object for reward_fn: {e}")
-        
+        #stop token
+        reward_fn.stop_token=config.actor_rollout_ref.rollout.stop[0]
```
Comment (James0618) [inline verl/trainer/main_ppo.py:186]: 这个.stop没看到定义呀

**Review by James0618 — COMMENTED** [review]
还需要在格式规范和实现位置上做一些调整

**Review by James0618 — APPROVED** [review]
Nice work!!!
